File: apps/worker-rs/src/checkers/redis.rs
use std::time::Duration;

use redis::AsyncCommands;
use url::Url;

use crate::types::{CheckResult, WorkerMonitor};

const RETRY_DELAY_MS: u64 = 1_500;

pub async fn check_redis(monitor: &WorkerMonitor) -> CheckResult {
    let first = perform_redis_check(monitor).await;
    if first.status != "down" {
        return first;
    }

    tokio::time::sleep(Duration::from_millis(RETRY_DELAY_MS)).await;
    perform_redis_check(monitor).await
}

async fn perform_redis_check(monitor: &WorkerMonitor) -> CheckResult {
    let started = std::time::Instant::now();
    let Some(url) = monitor.url.as_deref() else {
        return CheckResult {
            status: "down".to_string(),
            latency: 0,
            status_code: None,
            message: Some("No connection URL configured".to_string()),
            response_headers: None,
            response_body: None,
        };
    };

    let timeout = Duration::from_secs(monitor.timeout.max(1) as u64);
    let work = async {
        let client = redis::Client::open(url)?;
        let mut connection = client.get_multiplexed_async_connection().await?;

        if let Some(command_text) = monitor
            .body
            .as_deref()
            .map(str::trim)
            .filter(|value| !value.is_empty())
        {
            let parts = command_text.split_whitespace().collect::<Vec<_>>();
            let mut command = redis::cmd(parts[0]);
            for argument in parts.iter().skip(1) {
                command.arg(argument);
            }
            let _: redis::Value = command.query_async(&mut connection).await?;
        } else {
            let _: String = connection.ping().await?;
        }

        Ok::<(), redis::RedisError>(())
    };

    let latency = started.elapsed().as_millis() as i32;
    match tokio::time::timeout(timeout, work).await {
        Ok(Ok(())) => CheckResult {
            status: "up".to_string(),
            latency,
            status_code: None,
            message: None,
            response_headers: None,
            response_body: None,
        },
        Ok(Err(error)) => CheckResult {
            status: "down".to_string(),
            latency,
            status_code: None,
            message: Some(format_error(&error.to_string(), url)),
            response_headers: None,
            response_body: None,
        },
        Err(_) => CheckResult {
            status: "down".to_string(),
            latency,
            status_code: None,
            message: Some("Connection timed out".to_string()),
            response_headers: None,
            response_body: None,
        },
    }
}

fn format_error(error: &str, url: &str) -> String {
    error.replace(url, &strip_credentials(url))
}

fn strip_credentials(url: &str) -> String {
    match Url::parse(url) {
        Ok(mut parsed) => {
            if !parsed.username().is_empty() || parsed.password().is_some() {
                let _ = parsed.set_username("***");
                let _ = parsed.set_password(Some("***"));
            }
            parsed.to_string()
        }
        Err(_) => "[invalid url]".to_string(),
    }
}

File: apps/worker/src/checkers/redis.ts
import type { Monitor } from "@unstatus/db";
import { RedisClient } from "bun";
import { assertPublicUrl } from "./egress.js";

const RETRY_DELAY_MS = 1500;

function stripCredentials(url: string) {
  try {
    const parsed = new URL(url);
    if (parsed.username || parsed.password) {
      parsed.username = "***";
      parsed.password = "***";
    }
    return parsed.toString();
  } catch {
    return "[invalid url]";
  }
}

function formatError(e: unknown, url: string): string {
  if (!(e instanceof Error)) return "Unknown error";
  // Strip any credentials that might appear in the error message
  return e.message.replace(url, stripCredentials(url));
}

async function performRedisCheck(monitor: Monitor) {
  const start = performance.now();

  if (!monitor.url) {
    return {
      status: "down" as const,
      latency: 0,
      statusCode: null,
      message: "No connection URL configured",
      responseHeaders: null,
      responseBody: null,
    };
  }

  let client: RedisClient | null = null;

  try {
    const result = await Promise.race([
      (async () => {
        await assertPublicUrl(monitor.url!, ["redis:", "rediss:"]);
        client = new RedisClient(monitor.url!);
        await client.connect();

        if (monitor.body) {
          const parts = monitor.body.trim().split(/\s+/);
          await client.send(parts[0], ...parts.slice(1));
        } else {
          await client.send("PING");
        }

        return { ok: true };
      })(),
      new Promise<never>((_, reject) =>
        setTimeout(() => reject(new Error("Connection timed out")), monitor.timeout * 1000),
      ),
    ]);

    return {
      status: result.ok ? "up" : "down",
      latency: Math.round(performance.now() - start),
      statusCode: null,
      message: null,
      responseHeaders: null,
      responseBody: null,
    };
  } catch (e) {
    return {
      status: "down" as const,
      latency: Math.round(performance.now() - start),
      statusCode: null,
      message: formatError(e, monitor.url),
      responseHeaders: null,
      responseBody: null,
    };
  } finally {
    try { client?.close(); } catch {}
  }
}

export async function checkRedis(monitor: Monitor) {
  const first = await performRedisCheck(monitor);
  if (first.status !== "down") return first;

  await new Promise((r) => setTimeout(r, RETRY_DELAY_MS));
  return performRedisCheck(monitor);
}
